File: python/aoc_2023/day06.py
from dataclasses import dataclass


@dataclass
class Race:
    time: int
    distance_to_beat: int

    def run(self) -> int:
        answer = 0
        for i in range(self.time):
            run_time = self.time - i
            distance = i * run_time
            if distance > self.distance_to_beat:
                answer += 1
        return answer

File: src/rust/day06.rs
#[derive(Debug)]
pub struct Race {
    duration: usize,
    distance_to_beat: usize,
    pub answer: u32,
}

impl Race {
    pub fn new(duration: usize, distance_to_beat: usize) -> Self {
        Self {
            duration,
            distance_to_beat,
            answer: 0,
        }
    }

    pub fn run(&mut self) {
        for i in 1..=self.duration {
            let run_time = self.duration - i;
            let distance = i * run_time;
            if distance > self.distance_to_beat {
                self.answer += 1;
            }
        }
    }
}
